### ISSUE TITLE
Highlight unsafe operations inside `unsafe` blocks

### ISSUE THREAD

**Issue by matklad** [intellij-rust/intellij-rust#3013]
Only some operations are actually unsafe inside an unsafe block (calls to usafe fn's, raw pointer dereference). It would be cool to highlight such operations!

See https://github.com/rust-analyzer/rust-analyzer/issues/190 for details.

I personally think this would be a rather important thing to add for folks who work with unsafe code a lot.

**Comment by gnzlbg** [comment]
There are only 4 things that `unsafe {}` blocks allow one to do ([TRPL 2nd Ed 19.1 Unsafe Superpowers](https://doc.rust-lang.org/book/second-edition/ch19-01-unsafe-rust.html#unsafe-superpowers)): 

> * Dereference a raw pointer
> * Call an unsafe function or method
> * Access or modify a mutable static variable
> * Implement an unsafe trait

Only the first three are relevant for this issue because implementing an unsafe trait requires an `unsafe impl` independently of whether this `impl` is added inside an `unsafe {}` block or not.

**Comment by garyttierney** [comment]
The commit referenced above (https://github.com/garyttierney/intellij-rust/commit/dc17f4fd99dc510f98ea89b1c486b98f5c6a091f) is a WIP implementation of this.  It re-uses the same style as Kotlin smart casts, though I'm not convinced this is the best visual representation of "unsafety".  

![screenshot from 2018-11-20 21-50-30](https://user-images.githubusercontent.com/1180094/48805140-7f8be600-ed0e-11e8-96ec-16405dee258c.png)

**Comment by gnzlbg** [comment]
This is so cool.

**Comment by vlad20012** [comment]
@garyttierney Maybe we shouldn't highlight a full function call and highlight the only name instead?
Also let's discuss possible highlighting style. I'm going to use smartcast-like highlighting for a stuff like #1815. But maybe it's not a problem to use the same style for unsafe

**Comment by garyttierney** [comment]
That's a cool feature!  And yes, that makes sense.  That lets us disambiguate in the following case (which my implementation doesn't do at present):

```rust
extern "C" {
    static TEST: i32;
}

unsafe fn test(_: i32, _: i32) {}
fn test2() -> i32 { 0 }

fn main() {
    test(test2(), TEST);
    ---- -------  ^^^^ unsafe 
    |       |
} unsafe  safe
```

I do think unsafe annotations should have a different style to the type of annotations you've referenced in #1815, but not quite sure what that would be.  A yellow warning style background seems more appropriate, but not quite correct.  It is something that can be easily resolved and is customizable by users and themes, so I can get this branch in a state that's ready for review first and we can bikeshed the aesthetic later :)

**Comment by vlad20012** [comment]
> so I can get this branch in a state that's ready for review first and we can bikeshed the aesthetic later

Of course yes! We can even merge it and then think about an appropriate style =) I just want to start the discussion. Maybe @amakeev can suggest something about the style

**Comment by gnzlbg** [comment]
There are certain colors that are often used in society and nature to convey "danger" and a lot of literature on those, e.g., (just one of many) [National Agricultural Safety Database Color Coding](http://nasdonline.org/131/d001678/color-coding.html) : 

> The Standard Color-Code System:
>
> * RED - Denotes fire safety equipment and safety containers for flammables. Identifies emergency devices (emergency shut-off switches, stop bar, buttons).
> * ORANGE - Be aware of machinery or equipment that can cut, crush, shock or cause other injury
> * YELLOW - Cautions against physical dangers (slipping, tripping, falling, caught-between and striking-against hazards).
> * GREEN - Locates first-aid equipment.
> * BLUE - Cautions against the use or movement of equipment being repaired or the starting of equipment.
> * MAGENTA AND YELLOW or BLACK AND YELLOW - Warns of radiation hazards.
> * BLACK, WHITE OR A COMBINATION - Controls and designates traffic movement, marks aisle, housekeeping areas and similar areas.

mentions that colors like orange, yellow, and magenta are used to warn about dangers and hazards. OTOH colors like green warn about helpful stuff in dangerous situations like first-aid equipment or exit doors, etc. 

One might want to take inspiration from these to construct styles that are readable for the different themes, but still transmit the appropriate feeling.

**Comment by mibac138** [comment]
Is there anything left to be done in this issue or did #3080 fix it?

**Comment by gnzlbg** [comment]
It appears that there are no tests for "implement an `unsafe` Trait" in that PR. cc @garyttierney

**Comment by ortem** [comment]
@mibac138 @gnzlbg  Yes, there are two other cases of unsafe operation. I've updated the issue description.
